### PR TITLE
wayland: Don't wait on event queues twice

### LINKED ISSUE #6204: SDL_WaitEvent not getting SDL_USEREVENT when running with wayland + libdecor
After this commit:

https://github.com/libsdl-org/SDL/pull/5703

some programs using `SDL_WaitEvent` stopped working when run with wayland with libdecor support.

As in `Wayland_WaitEventTimeout` the call to `libdecor_dispatch` will flush all pending events the function will block at `SDL_IOReady(WAYLAND_wl_display_get_fd(d->display)...)` and the user events won't be processed.

One example of application failing is scrcpy under latest Fedora.

here is a test case showing the problem in different scenarios:

https://gist.github.com/fernandoherreradelasheras/ec2f3cbf57f38ba876148493ca1f7cf6

- 0: start pushing events without the wayland window shown -> No events at all will be processed
- 1 and 2: push events after window is shown -> Some events will be properly processed but at some point WaitEvent will block again for a while
- 3: using SDL_PollEvent -> works fine.

### PR DESCRIPTION
**Pull request by Kontrabant**
libdecor_dispatch() needs to be called, as libdecor plugins might do some required internal processing within, however care must be taken to avoid double-blocking in the case of a timeout, which can occur if libdecor_dispatch() and the SDL event processing both work on the main Wayland queue. Additionally, assumptions that libdecor will always dispatch the main queue or not process zero-length queues (can occur if a wait is interrupted by the application queueing an event) should not be made, as this behavior is outside the control of SDL and can change.

SDL handles polling for Wayland events and then calls libdecor to do its internal processing and dispatch. If libdecor operates on the main queue, it will dispatch the queued events and the additional wl_display_dispatch_pending() call will be a NOP. If a libdecor plugin uses its own, separate queue, then the wl_display_dispatch_pending() call will ensure that the main queue is always dispatched.

Fixes #6204 
Doesn't completely fix, but improves #6074 (`SDL_WaitEvent()` doesn't indefinitely block even when events are available and never returns 0)

### CODE REVIEW

**Review by flibitijibibo — APPROVED** [review]
Makes sense to me, but I won't be able to test until Monday - if someone else can verify, this is good to go.

**Comment by Kontrabant** [comment]
Not yet. After thinking about it I ran into a few edge cases where this can fail, hence why I changed it to draft.

`libdecor_dispatch()` is a very annoying function that doesn't really work for our purposes as it doesn't distinguish between a return value of 0 for timeout and a return value of 0 for when it was woken up but there were zero events in the queue to process, as can happen when an application manually queues an SDL event.

The current version works, but looking into libdecor, it relies on some very specific behavior in it's dispatch implementation that could break SDL if it is ever changed.  Likewise, this also assumes that libdecor plugins will never use their own queues, which is a theoretical problem as well.

I have a more robust, albeit less elegant, version of this that doesn't rely on behavior outside of our control.  I'll have it up tomorrow.

**Comment by Kontrabant** [comment]
New version that doesn't depend on any specific, internal libdecor behavior or the ambiguous return value of `libdecor_dispatch()`.

File: src/video/wayland/SDL_waylandevents.c [hunk 1]
```diff
@@ -289,6 +289,24 @@ Wayland_SendWakeupEvent(_THIS, SDL_Window *window)
...
+static int
+dispatch_queued_events(SDL_VideoData *viddata)
+{
+    int ret;
+
+#ifdef HAVE_LIBDECOR_H
+    if (viddata->shell.libdecor) {
+        ret = libdecor_dispatch(viddata->shell.libdecor, 0);
+        if (ret < 0) {
+            return ret;
```
Comment (slouken) [inline src/video/wayland/SDL_waylandevents.c:301]: Do we want to fail here and not call `WAYLAND_wl_display_dispatch_pending()` if libdecor has an issue?

Comment (Kontrabant) [reply src/video/wayland/SDL_waylandevents.c:301]: The assumption was that a dispatch error is something serious like a disconnection, but that should be caught by SDL as well when it tries to dispatch the queue immediately after this. This check can probably go.

Comment (flibitijibibo) [reply src/video/wayland/SDL_waylandevents.c:301]: If not a check, probably a note for #6213 if we need to care about libdecor's involvement with reconnect support.
